_doc/server_config.md
# サーバの設定

## 目次

- [サーバプログラムのビルド](#サーバプログラムのビルド)
- [データベースの構築](#データベースの構築)
- [サーバ設定ファイル](#サーバ設定ファイル)

## サーバプログラムのビルド

WSNet2のサーバはGoで書かれたプログラムです。
ビルドするにはGo（1.18以降）のほか、GNU MakeとUnix風シェルが必要です。

[`server`ディレクトリ](../server)にて、`make`コマンドを実行すると、

`bin`ディレクトリに次の実行ファイルが作られます。
WSNet2の機能として必要なのはLobby、Game、Hubの3種類のサーバです。

- **wsnet2-lobby**: Lobbyサーバ
- **wsnet2-game**: Gameサーバ
- **wsnet2-hub**: Hubサーバ
- **wsnet2-bot**: 負荷試験やシナリオ試験用のbotクライアント
- **wsnet2-tool**: サーバや部屋の情報を閲覧するコマンドラインツール

## データベースの構築

MySQL互換RDBMSを利用します。

必要なテーブルは[`sql/10-schema.sql`](../server/sql/10-schema.sql)に定義されています。

- **app**: 登録アプリ識別子と鍵
- **game_server**: Gameサーバの接続情報と状態
- **hub_server**: Hubサーバの接続情報と状態
- **room**: 稼働中の部屋
- **hub**: 稼働中の観戦用部屋
- **room_history**: 終了した部屋

`app`テーブルにAppIDとKeyを登録します。この情報はゲームAPIサーバと共有するものでユーザ認証に使われます。
詳しくは[WSNet2のユーザ認証](user_auth.md#鍵の事前交換)を参照して下さい。

## サーバ設定ファイル

サーバプログラム（wsnet2-lobby、wsnet2-game、wsnet2-hub）の起動時には、
コマンドライン引数としてTOML形式の設定ファイルを指定します。

サーバ種別ごと別テーブルに分けているので、同一のファイルを全種類のサーバに使えます。

また、GameとHubの`hostname`、`public_name`、`grpc_port`、`websocket_port`は次の環境変数で上書きできます。
複数台構成ではホスト名を環境変数で指定することで、設定ファイルを共通にできます。

- `WSNET2_GAME_HOSTNAME`
- `WSNET2_GAME_PUBLICNAME`
- `WSNET2_GAME_GRPCPORT`
- `WSNET2_GAME_WSPORT`

### ファイル内容

```toml
#
# RDBMSに関する設定
#
[Database]

# 接続先
host = "wsnet2-db"
port = 3306
dbname = "wsnet2"

# 接続ユーザ
user = "wsnet"
password = "password"
# user, passwordの代わりに、authfileで指定することもできます
authfile = "/path/to/authfile" # "user:password" が書かれたファイル

conn_max_lifetime = "3m" # 接続を再利用できる最大時間（デフォルト: 3min）

#
# Lobbyサーバに関する設定
#
[Lobby]
hostname = "" # 指定した場合、ホスト名がマッチしたHTTPリクエストのみ受け付ける
net = "tcp"       # ソケット種別。"unix": UNIXドメインソケット
unixpath = ""     # net="unix"のときのunixドメインソケットのパス
port = 8080       # net="tcp"のときのポート番号
pprof_prot = 3080 # pprofの待受けポート番号

valid_heartbeat = "5s" # Game,Hubの最終HeartBeat時刻の有効期間（デフォルト:5s）
authdata_expire = "1m" # AuthDataの有効期間（デフォルト:1m）
api_timeout = "5s"     # LobbyAPIの内部タイムアウト時間（デフォルト:5s）
db_max_conns = 0       # 最大DB接続数
hub_max_watchers = 10000 # Hubサーバの最大収容観戦者数

# ログ設定
loglevel = 5 # 基本ログレベル（デフォルト:2）
logconf.log_stdout_level = 4       # stdoutのログレベル
logconf.log_stdout_console = false # stdoutのログフォーマットを開発用にする
# ログファイル出力とローテーションの設定
logconf.log_path = ""         # 空ならファイルに出力しない
logconf.log_max_size = 500
logconf.log_max_backups = 0
logconf.log_max_age = 0
logconf.log_compress = false

#
# Gameサーバの設定
#
[Game]
hostname = "wsnet2-game"                # ローカルホスト名（Lobby, Hubからのアクセス）
public_name = "wsnet2-game.example.com" # 公開ホスト名（クライアントからのアクセス）
grpc_port = 19000                       # gRPC待受けポート（Lobby, Hubからのアクセス）
websocket_port = 8000                   # WebSocket待受けポート（クライアント、Hubからのアクセス）
pprof_port = 3000
# WebSocket接続にTLSを使用する場合の設定
# 空ならTLSを使用しない
tls_cert = "/path/to/cert_file"
tls_key = "/path/to/key_file"

retry_count = 5        # ユニークな部屋番号生成のリトライ回数（デフォルト:5）
max_room_num = 999999  # 部屋番号の最大値。3桁に制限したいときは 999 とする
max_rooms = 1000       # 最大部屋数（デフォルト：1000）
max_clients = 5000     # 最大クライアント数（デフォルト：5000）
db_max_conns = 0       # 最大DB接続数
heartbeat_interval = "2s" # HeartBeat時刻更新間隔。{Lobby,Hub}.valid_heartbeatより短くする。
# 部屋の初期値
default_max_players = 10 # 部屋あたりの最大プレイヤー数（デフォルト:10）
default_deadline = 5     # クライアントタイムアウト判定時間（秒; デフォルト:5）
default_loglevel = 2     # 部屋のログレベル
# client設定
client.event_buf_size = 128     # イベント再送バッファ数（デフォルト:128）
client.wait_after_close = "30s" # 部屋終了後の再接続データ再送可能時間（デフォルト:30s）
auth_key_len = 32               # 接続のユーザ認証用の鍵のサイズ

# ログ設定（Lobbyと同じ）
loglevel = 2
logconf.log_stdout_level = 4
logconf.log_stdout_console = false
logconf.log_path = ""
logconf.log_max_size = 500
logconf.log_max_backups = 0
logconf.log_max_age = 0
logconf.log_compress = false

#
# Hubサーバの設定
#
[Hub]
# 基本的にGameと同じ
hostname = "wsnet2-hub"
public_name = "wsnet2-hub.example.com"
grpc_port = 19010
websocket_port = 8010
pprof_port = 3010
tls_cert = "/path/to/cert/file"
tls_key = "/path/to/key/file"
max_clients = 5000
default_loglevel = 2
valid_heartbeat = "5s"     # Gameの最終HeartBeat時刻の有効期間（デフォルト:5s）
heartbeat_interval = "2s"
nodecount_interval = "1s"  # Hubを経由している観戦者数の同期間隔（デフォルト:1s）
db_max_conns = 0
client.event_buf_size = 128
client.wait_after_close = "30s"
auth_key_len = 32
loglevel = 2
logconf.log_stdout_level = 4
logconf.log_stdout_console = false
logconf.log_path = ""
logconf.log_max_size = 500
logconf.log_max_backups = 0
logconf.log_max_age = 0
logconf.log_compress = false
```


_doc/wsnet2settings.md
# WSNet2Settings クラス

- [概要](#概要)
- [設定項目](#設定項目)
  - [EvPoolSize](#EvPoolSize)
  - [EvBufInitialSize](#EvBufInitialSize)
  - [MsgPoolSize](#MsgPoolSize)
  - [MsgBufInitialSize](#MsgBufInitialSize)
  - [MaxReconnection](#MaxReconnection)
  - [ConnectTimeoutMilliSec](#ConnectTimeoutMilliSec)
  - [RetryIntervalMilliSec](#RetryIntervalMilliSec)
  - [MaxPingIntervalMilliSec](#MaxPingIntervalMilliSec)
  - [MinPingIntervalMilliSec](#MinPingIntervalMilliSec)

## 概要

WSNet2をプロジェクトに合わせて調整するための設定を[`WSNet2.WSNet2Settings`](../wsnet2-unity/Assets/WSNet2/Scripts/Core/WSNet2Settings.cs)クラスを通して行えます。

## 設定項目

### EvPoolSize

- 保持できるEventの数
- 初期値: 16 個

入室時の`onSuccess`コールバック中や`room.Pause()`で一時停止している間に受信したEventを保持するためのプールサイズです。
自動拡張されないので、十分な量を設定してください。

### EvBufInitialSize

- 各Eventのバッファサイズの初期値
- 初期値: 256 byte

各イベントを保持するバッファサイズの初期値です。
`EvPoolSize`個のバッファが確保されます。
各バッファは必要に応じて自動的に拡張されます。

### MsgPoolSize

- 保持できるMsgの数
- 初期値: 16 個

送信待ちメッセージを保持するためのプールのサイズです。
回線切断による再接続中のメッセージをプールするだけでなく、再接続時の不達メッセージの再送にも使われます。

### MsgBufInitialSize

- 各Msgのバッファサイズの初期値
- 初期値: 256 byte

各メッセージを保持するバッファサイズの初期値です。
`MsgPoolSize`個のバッファが確保されます。
各バッファは必要に応じて自動的に拡張されます。

### MaxReconnection

- 最大連続再接続試行回数
- 初期値: 5 回

websocket再接続試行回数の上限です。
この回数連続して接続に失敗したら再試行を打ち切り、部屋から切断したことになります。
再接続に成功したら失敗回数のカウントは0に戻ります。

### ConnectTimeoutMillSec

- 接続タイムアウト
- 初期値: 5000 ミリ秒 (=5秒)

websocket接続のタイムアウト時間です。
接続要求に対してこの時間応答が無かった場合、その接続は失敗として扱われます。

### RetryIntervalMilliSec

- 再接続インターバル
- 初期値: 1000 ミリ秒 (=1秒)

接続要求してから、次の再接続を要求するまでの最低間隔です。
この時間以上接続していたり、タイムアウト時間がこの時間より長い場合は、即時に再接続が試行されます。

### MaxPingIntervalMilliSec

- 最大Ping間隔
- 初期値: 10000 ミリ秒 (=10秒)

Ping間隔は部屋に設定された[`ClientDeadline`](room.md#clientdeadline)に応じて自動計算されますが、
この時間より長くなることはありません。

WSNet2では、Pingメッセージに対応するPongイベントによって観戦者数や各Playerの最終Msg受信時刻を同期するため、
あまり長い間隔にしてしまうとこれらの情報のリアルタイム性が損なわれてしまいます。

### MinPingIntervalMilliSec

- 最小Ping間隔
- 初期値: 1000 ミリ秒 (=1秒)

Ping間隔はRoomの`ClientDeadline`に応じて自動計算されますが、通信頻度が過剰にならないようにするため、この時間より短くなることはありません。
逆に言えば、`ClientDeadline`がこの値より短い場合、タイムアウトしないためのPingが機能しなくなります。


server/config/config.go
package config

import (
	"fmt"
	"os"
	"path"
	"strconv"
	"strings"
	"time"

	"github.com/pelletier/go-toml"
	"golang.org/x/xerrors"
)

type Config struct {
	Db    DbConf `toml:"Database"`
	Game  GameConf
	Hub   HubConf
	Lobby LobbyConf
}

type LogConf struct {
	// stdout をローカル開発用のフォーマットにする
	LogStdoutConsole bool `toml:"log_stdout_console"`
	// stdout のログレベル設定
	LogStdoutLevel uint32 `toml:"log_stdout_level"`

	// ローテーション設定
	// https://github.com/natefinch/lumberjack#type-logger
	LogPath       string `toml:"log_path"`
	LogMaxSize    int    `toml:"log_max_size"`
	LogMaxBackups int    `toml:"log_max_backups"`
	LogMaxAge     int    `toml:"log_max_age"`
	LogCompress   bool   `toml:"log_compress"`
}

type DbConf struct {
	Host            string
	Port            int
	DBName          string
	AuthFile        string
	User            string
	Password        string
	ConnMaxLifetime Duration `toml:"conn_max_lifetime"`
}

type GameConf struct {
	// Hostname : Lobbyなどからのアクセス名. see Load()
	Hostname string
	// PublicName : クライアントからのアクセス名. see Load()
	PublicName string `toml:"public_name"`

	GRPCPort      int `toml:"grpc_port"`
	WebsocketPort int `toml:"websocket_port"`
	PprofPort     int `toml:"pprof_port"`

	TLSCert string `toml:"tls_cert"`
	TLSKey  string `toml:"tls_key"`

	RetryCount int `toml:"retry_count"`
	// MaxRoomNum : 部屋番号最大値
	MaxRoomNum int `toml:"max_room_num"`

	// MaxRooms : 最大部屋数
	MaxRooms int `toml:"max_rooms"`
	// MaxClients : サーバ当たりの最大クライアント数
	MaxClients int `toml:"max_clients"`

	DefaultMaxPlayers uint32 `toml:"default_max_players"`
	DefaultDeadline   uint32 `toml:"default_deadline"`
	DefaultLoglevel   uint32 `toml:"default_loglevel"`

	HeartBeatInterval Duration `toml:"heartbeat_interval"`

	DbMaxConns int `toml:"db_max_conns"`

	ClientConf
	LogConf
}

type HubConf struct {
	// Hostname : Lobbyなどからのアクセス名. see Load()
	Hostname string
	// PublicName : クライアントからのアクセス名. see Load()
	PublicName string `toml:"public_name"`

	GRPCPort      int `toml:"grpc_port"`
	WebsocketPort int `toml:"websocket_port"`
	PprofPort     int `toml:"pprof_port"`

	TLSCert string `toml:"tls_cert"`
	TLSKey  string `toml:"tls_key"`

	// MaxClients : サーバ当たりの最大クライアント数
	MaxClients int `toml:"max_clients"`

	DefaultLoglevel uint32 `toml:"default_loglevel"`

	// ValidHeartBeat : HeartBeatの有効期間（GameCacheで使用）
	ValidHeartBeat Duration `toml:"valid_heartbeat"`

	HeartBeatInterval Duration `toml:"heartbeat_interval"`
	NodeCountInterval Duration `toml:"nodecount_interval"`

	DbMaxConns int `toml:"db_max_conns"`

	ClientConf
	LogConf
}

type ClientConf struct {
	EventBufSize int `toml:"event_buf_size"`

	// 部屋が終了した後で再接続が来た時もバッファに残ったデータを送信できるので一定時間残す
	WaitAfterClose Duration `toml:"wait_after_close"`

	AuthKeyLen int `toml:"auth_key_len"`
}

type LobbyConf struct {
	Hostname  string
	UnixPath  string
	Net       string
	Port      int
	PprofPort int `toml:"pprof_port"`

	Loglevel uint32 `toml:"loglevel"`

	// ValidHeartBeat : HeartBeatの有効期間
	ValidHeartBeat Duration `toml:"valid_heartbeat"`

	AuthDataExpire Duration `toml:"authdata_expire"`

	ApiTimeout Duration `toml:"api_timeout"`

	HubMaxWatchers int `toml:"hub_max_watchers"`

	DbMaxConns int `toml:"db_max_conns"`

	LogConf
}

type Duration time.Duration

func (d *Duration) UnmarshalText(text []byte) error {
	td, err := time.ParseDuration(string(text))
	*d = Duration(td)
	return err
}

// Load : tomlファイルから読み込む
//
// 次の環境変数はtomlより優先される.
// - WSNET2_GAME_HOSTNAME:   Config.{Game,Hub}.Hostname
// - WSNET2_GAME_PUBLICNAME: Config.{Game,Hub}.PublicName
// - WSNET2_GAME_GRPCPORT:   Config.{Game,Hub}.GRPCPort
// - WSNET2_GAME_WSPORT:     Config.{Game,Hub}.WebsocketPort
//
func Load(conffile string) (*Config, error) {
	hostname, _ := os.Hostname()
	if hostname == "" {
		hostname = "localhost"
	}

	c := &Config{
		// set default values before decode file.
		Db: DbConf{
			ConnMaxLifetime: Duration(3 * time.Minute),
		},
		Game: GameConf{
			Hostname:   hostname,
			PublicName: hostname,

			RetryCount: 5,
			MaxRoomNum: 999999,

			MaxRooms:   1000,
			MaxClients: 5000,

			DefaultMaxPlayers: 10,
			DefaultDeadline:   5,
			DefaultLoglevel:   2,

			HeartBeatInterval: Duration(2 * time.Second),

			DbMaxConns: 0,

			ClientConf: ClientConf{
				EventBufSize:   128,
				WaitAfterClose: Duration(30 * time.Second),
				AuthKeyLen:     32,
			},

			LogConf: LogConf{
				LogStdoutLevel: 4,
				LogPath:        "/var/log/wsnet2/wsnet2-game.log",
				LogMaxSize:     500,
				LogMaxBackups:  0,
				LogMaxAge:      0,
				LogCompress:    false,
			},
		},
		Hub: HubConf{
			Hostname:   hostname,
			PublicName: hostname,

			MaxClients: 5000,

			DefaultLoglevel: 2,

			ValidHeartBeat: Duration(5 * time.Second),

			HeartBeatInterval: Duration(2 * time.Second),
			NodeCountInterval: Duration(1 * time.Second),

			DbMaxConns: 0,

			ClientConf: ClientConf{
				EventBufSize:   128,
				WaitAfterClose: Duration(30 * time.Second),
				AuthKeyLen:     32,
			},

			LogConf: LogConf{
				LogStdoutLevel: 4,
				LogPath:        "/var/log/wsnet2/wsnet2-hub.log",
				LogMaxSize:     500,
				LogMaxBackups:  0,
				LogMaxAge:      0,
				LogCompress:    false,
			},
		},
		Lobby: LobbyConf{
			ValidHeartBeat: Duration(5 * time.Second),
			Loglevel:       2,
			AuthDataExpire: Duration(time.Minute),
			ApiTimeout:     Duration(5 * time.Second),
			HubMaxWatchers: 10000,

			DbMaxConns: 0,

			LogConf: LogConf{
				LogStdoutLevel: 4,
				LogPath:        "/var/log/wsnet2/wsnet2-lobby.log",
				LogMaxSize:     500,
				LogMaxBackups:  0,
				LogMaxAge:      0,
				LogCompress:    false,
			},
		},
	}

	confBytes, err := os.ReadFile(conffile)
	if err != nil {
		return nil, err
	}

	err = toml.Unmarshal(confBytes, c)
	if err != nil {
		return nil, err
	}

	err = c.Db.loadAuthfile(conffile)
	if err != nil {
		return nil, err
	}

	c.applyEnvVar()

	return c, nil
}

func (db *DbConf) loadAuthfile(conffile string) error {
	if db.AuthFile == "" {
		return nil
	}
	authfile := db.AuthFile
	if authfile[0] != '/' {
		authfile = path.Join(path.Dir(conffile), authfile)
	}
	content, err := os.ReadFile(authfile)
	if err != nil {
		return err
	}
	ss := strings.SplitN(strings.TrimSpace(string(content)), ":", 2)
	if len(ss) != 2 {
		return xerrors.Errorf("Db authfile format error: %q", string(content))
	}

	db.User = ss[0]
	db.Password = ss[1]
	return nil
}

func (db *DbConf) DSN() string {
	user := db.User
	if db.Password != "" {
		user = fmt.Sprintf("%s:%s", db.User, db.Password)
	}
	return fmt.Sprintf("%s@tcp(%s:%d)/%s?parseTime=true", user, db.Host, db.Port, db.DBName)
}

// applyEnvVar : 環境変数で上書きする
func (c *Config) applyEnvVar() {
	if v := os.Getenv("WSNET2_GAME_HOSTNAME"); v != "" {
		c.Game.Hostname = v
		c.Hub.Hostname = v
	}
	if v := os.Getenv("WSNET2_GAME_PUBLICNAME"); v != "" {
		c.Game.PublicName = v
		c.Hub.PublicName = v
	}
	if v, err := strconv.Atoi(os.Getenv("WSNET2_GAME_WSPORT")); err == nil {
		c.Game.WebsocketPort = v
		c.Hub.WebsocketPort = v
	}
	if v, err := strconv.Atoi(os.Getenv("WSNET2_GAME_GRPCPORT")); err == nil {
		c.Game.GRPCPort = v
		c.Hub.GRPCPort = v
	}
}
